File: applications/luci-app-unbound/luasrc/model/cbi/unbound/zones.lua
-- Copyright 2017 Eric Luehrsen <ericluehrsen@gmail.com>
-- Licensed to the public under the Apache License 2.0.

local m5, s5
local ztype, zones, servers, fallback, enabled

local fs = require "nixio.fs"
local ut = require "luci.util"
local sy = require "luci.sys"
local resolvfile = "/tmp/resolv.conf.auto"

m5 = Map("unbound")
s5 = m5:section(TypedSection, "zone", "Zones",
    translatef("This shows extended zones and more details can be "
    .. "changed in Files tab and <a href=\"%s\">Edit:UCI</a> subtab.",
    "/cgi-bin/luci/admin/services/unbound/files" ))

s5.addremove = false
s5.anonymous = true
s5.sortable = true
s5.template = "cbi/tblsection"

ztype = s5:option(DummyValue, "DummyType", translate("Type"))
ztype.rawhtml = true

zones = s5:option(DummyValue, "DummyZones", translate("Zones"))
zones.rawhtml = true

servers = s5:option(DummyValue, "DummyServers", translate("Servers"))
servers.rawhtml = true

fallback = s5:option(Flag, "fallback", translate("Fallback"))
fallback.rmempty = false

enabled = s5:option(Flag, "enabled", translate("Enable"))
enabled.rmempty = false


function ztype.cfgvalue(self, s)
    -- Format a meaninful tile for the Zone Type column
    local itxt = self.map:get(s, "zone_type")
    local itls = self.map:get(s, "tls_upstream")


    if itxt and itxt:match("forward") then
        if itls and (itls == "1") then
            return translate("Forward TLS")

        else
            return translate("Forward")
        end

    elseif itxt and itxt:match("stub") then
        return translate("Recurse")

    elseif itxt and itxt:match("auth") then
        return translate("AXFR")

    else
        return translate("Error")
    end
end


function zones.cfgvalue(self, s)
    -- Format a meaninful sentence for the Zones viewed column
    local xtxt, otxt
    local itxt = self.map:get(s, "zone_name")
    local itype = self.map:get(s, "zone_type")


    for xtxt in ut.imatch(itxt) do
        if (xtxt == ".") then
            -- zone_name lists
            xtxt = translate("(root)")
        end


        if otxt and (#otxt > 0) then
            otxt = otxt .. ", <var>%s</var>" % xtxt

        else
            otxt = "<var>%s</var>" % xtxt
        end
    end


    if itype and itype:match("forward") then
        -- from zone_type create a readable hint for the action
        otxt = translate("accept upstream results for ") .. otxt

    elseif itype and itype:match("stub") then
        otxt = translate("select recursion for ") .. otxt

    elseif itype and itype:match("auth") then
        otxt = translate("prefetch zone files for ") .. otxt

    else
        otxt = translate("unknown action for ") .. otxt
    end


    if otxt and (#otxt > 0) then
        return otxt

    else
        return "(empty)"
    end
end


function servers.cfgvalue(self, s)
    -- Format a meaninful sentence for the Servers (and URL) column
    local xtxt, otxt, rtxt, found
    local itxt = self.map:get(s, "server")
    local iurl = self.map:get(s, "url_dir")
    local itype = self.map:get(s, "zone_type")
    local itls = self.map:get(s, "tls_upstream")
    local iidx = self.map:get(s, "tls_index")
    local irslv = self.map:get(s, "resolv_conf")


    for xtxt in ut.imatch(itxt) do
        if otxt and (#otxt > 0) then
            -- bundle and make pretty the server list
            otxt = otxt .. ", <var>%s</var>" % xtxt

        else
            otxt = "<var>%s</var>" % xtxt
        end
    end


    if otxt and (#otxt > 0)
    and itls and (itls == "1")
    and iidx and (#iidx > 0) then
        -- show TLS certificate name index if provided
        otxt = translatef("use nameservers by <var>%s</var> at ", iidx) .. otxt

    elseif otxt and (#otxt > 0) then
        otxt = translate("use nameservers ") .. otxt
    end


    if iurl and (#iurl > 0) and itype and itype:match("auth") then
        if otxt and (#otxt > 0) then
            -- include optional URL filed for auth-zone: type
            otxt = otxt .. translatef(", and try <var>%s</var>", iurl)

        else
            otxt = translatef("download from <var>%s</var>", iurl)
        end
    end


    if irslv and (irslv == "1") and itype and itype:match("forward") then
        for xtxt in ut.imatch(fs.readfile(resolvfile)) do
            if xtxt:match("nameserver") then
                found = true

            elseif (found == true) then
                if rtxt and (#rtxt > 0) then
                    -- fetch name servers from resolv.conf
                    rtxt = rtxt .. ", <var>%s</var>" % xtxt

                else
                    rtxt = "<var>%s</var>" % xtxt
                end


                found = false
            end
        end


        if otxt and (#otxt > 0) and rtxt and (#rtxt > 0) then
            otxt = otxt
                .. translatef(", and <var>%s</var> entries ", resolvfile) .. rtxt

        elseif rtxt and (#rtxt > 0) then
            otxt = translatef("use <var>%s</var> nameservers ", resolvfile) .. rtxt
        end
    end


    if otxt and (#otxt > 0) then
        return otxt

    else
        return "(empty)"
    end
end


function m5.on_commit(self)
    if sy.init.enabled("unbound") then
        -- Restart Unbound with configuration
        sy.call("/etc/init.d/unbound restart >/dev/null 2>&1")

    else
        sy.call("/etc/init.d/unbound stop >/dev/null 2>&1")
    end
end


return m5

File: applications/luci-app-unbound/htdocs/luci-static/resources/view/unbound/zones.js
'use strict';
'require form';
'require fs';
'require rpc';
'require uci';
'require ui';
'require view';

const RESOLV_FILE = '/tmp/resolv.conf.d/resolv.conf.auto';
const LOGERR_CMD = "logread -e 'unbound.*error.*ssl library'";
const HELP_URL = 'https://github.com/openwrt/packages/blob/master/net/unbound/files/README.md';

let section_enabled = false;
let daemon_enabled = false;

const callRcInit = rpc.declare({
	object: 'rc',
	method: 'init',
	params: [ 'name', 'action' ],
	expect: { result: false }
});

return view.extend({
	load() {
		return Promise.all([
			uci.load('unbound'),
			fs.read(RESOLV_FILE).catch(() => ''),
			fs.exec(LOGERR_CMD).catch(() => ''),
		]).then(([_, resolv, logerr]) => ({
			resolvContent: resolv || '',
			logerr: logerr ? logerr.trim().slice(-250) : null,
		}));
	},

	render({resolvContent, logerr}) {
		const m = new form.Map('unbound');

		daemon_enabled = uci.get('unbound', 'ub_main', 'enabled');

		if (logerr) {
			ui.addTimeLimitedNotification(null, E('p', _('Note: SSL/TLS library is missing an API. Please review syslog. >> logread ... ') + logerr), 7000, 'warning');
		}

		const s = m.section(form.TableSection, 'zone', _('Zones'),
			_('Organize directed forward, stub, and authoritative zones' +
			' <a href=\"%s\" target=\"_blank\">(help)</a>.'.format(HELP_URL)));
		s.anonymous = false;
		s.addremove = true;
		s.sortable = true;

		s.extedit = L.url(`admin/services/unbound/zone-details/%s`);

		const ztype = s.option(form.DummyValue, 'zone_type', _('Type'));
		ztype.rawhtml = true;
		ztype.cfgvalue = function(section_id) {
			const type = uci.get('unbound', section_id, 'zone_type');
			const tls = uci.get('unbound', section_id, 'tls_upstream');
			if (type?.includes('forward')) return tls === '1' ? _('Forward TLS') : _('Forward');
			if (type?.includes('stub')) return _('Recurse');
			if (type?.includes('auth')) return _('AXFR');
			return _('Undefined');
		};

		const zones = s.option(form.DummyValue, 'zone_name', _('Zones'));
		zones.rawhtml = true;
		zones.cfgvalue = function(section_id) {
			const type = uci.get('unbound', section_id, 'zone_type');
			const raw = uci.get('unbound', section_id, 'zone_name');
			const list = Array.isArray(raw) ? raw : raw?.split(/\s+/) || [];
			if (list.length === 0) return '(empty)';

			const html = list.map(name =>
				`<var>${name === '.' ? _('(root)') : name}</var>`
			).join(', ');

			if (type?.includes('forward')) return _('accept upstream results for ') + html;
			if (type?.includes('stub')) return _('select recursion for ') + html;
			if (type?.includes('auth')) return _('prefetch zone files for ') + html;

			return _('unknown action for ') + html;
		};

		const servers = s.option(form.DummyValue, 'server', _('Servers'));
		servers.rawhtml = true;
		servers.cfgvalue = function(section_id) {
			const servers = uci.get('unbound', section_id, 'server');
			const url = uci.get('unbound', section_id, 'url_dir');
			const type = uci.get('unbound', section_id, 'zone_type');
			const tls = uci.get('unbound', section_id, 'tls_upstream');
			const tlsName = uci.get('unbound', section_id, 'tls_index');
			const useResolv = uci.get('unbound', section_id, 'resolv_conf');

			const list = Array.isArray(servers) ? servers : servers?.split(/\s+/) || [];
			let out = list.map(s => `<var>${s}</var>`).join(', ');

			if (out) out = _('use nameservers ') + out;

			if (tls === '1' && tlsName)
				out += _(' with default certificate for <var>%s</var>'.format(tlsName));

			if (type?.includes('auth') && url)
				out += _(', and try <var>%s</var>'.format(url));
			else if (!type?.includes('auth') && url)
				out += _('download from <var>%s</var>'.format(url));

			if (type?.includes('forward') && useResolv === '1') {
				const lines = resolvContent.split('\n');
				const resolvNameservers = [];
				for (let i = 0; i < lines.length; i++) {
					if (lines[i].match(/^\s*nameserver\b/)) {
						resolvNameservers.push(`<var>${lines[i].match(/^\s*nameserver\b(.*)$/)[1]}</var>`);
					}
				}
				if (!out && resolvNameservers.length > 0)
					out = _('use <var>%s</var> nameservers '.format(RESOLV_FILE)) + resolvNameservers.join(', ');
				else if (resolvNameservers.length > 0)
					out += _(', and <var>%s</var> entries '.format(RESOLV_FILE)) + resolvNameservers.join(', ');
			}

			return out || '(empty)';
		};

		s.option(form.Flag, 'fallback', _('Fallback')).rmempty = false;
		const ena = s.option(form.Flag, 'enabled', _('Enable'))
		ena.rmempty = false;
		ena.load = (section_id) => { 
			let _temp = uci.get('unbound', section_id, 'enabled');
			if (_temp === '1') section_enabled = true;
		}

		return m.render();
	},

	handleSaveApply(ev, mode) {
		const cmd = daemon_enabled ? 'reload' : 'stop';

		const Fn = L.bind(() => {
			if (section_enabled) callRcInit('unbound', cmd);
			document.removeEventListener('uci-applied', Fn);
		});
		document.addEventListener('uci-applied', Fn);
		this.super('handleSaveApply', [ev, mode]);
	}
});
